File: 프로그래머스/unrated/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    if n == 1:
        return [[1]]
    direction = {'r': [0, 1], 'd': [1, 0], 'l': [0, -1], 'u': [-1, 0]}
    
    answer = [[0]*n for i in range(n)]
    x = 0
    y = 0
    dir = 'r'
    
    for i in range(1, n**2+1):
        answer[x][y] = i
        if dir == 'r':
            if y == n-1 or answer[x][y+1] != 0:
                dir = 'd'
        elif dir == 'd':
            if x == n-1 or answer[x+1][y] != 0:
                dir = 'l'
        elif dir == 'l':
            if y == 0 or answer[x][y-1] != 0:
                dir = 'u'
        elif dir == 'u':
            if x == n-1 or answer[x-1][y] != 0:
                dir = 'r'
        x += direction[dir][0]
        y += direction[dir][1]
                
    return answer

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {

    final int[][] DIRECT = { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };

    public int[][] solution(int n) {

        int[][] answer = new int[n][n];

        int row = 0, col = 0, value = 0, d = 0;

        while (value < n * n) {

            answer[row][col] = ++value;

            if (row + DIRECT[d][0] < 0 || n <= row + DIRECT[d][0] || col + DIRECT[d][1] < 0 || n <= col + DIRECT[d][1]
                    || answer[row + DIRECT[d][0]][col + DIRECT[d][1]] != 0) {
                d = (d + 1) % 4;
            }

            row += DIRECT[d][0];
            col += DIRECT[d][1];

        }

        return answer;
    }
}